archive/demos/shiny-drum-machine-audio.js

```diff
@@ -0,0 +1,307 @@
+/* eslint require-jsdoc: "off" */
+
+import {INSTRUMENTS, freeze, clone} from './shiny-drum-machine-data.js';
+
+const AudioContext = window.AudioContext ?? window.webkitAudioContext;
+
+const context = new AudioContext();
+
+const LOOP_LENGTH = 16;
+const BEATS_PER_FULL_NOTE = 4;
+const VOLUMES = freeze([0, 0.3, 1]);
+
+
+async function fetchAudio(url) {
+  const response = await fetch(new Request(url));
+  const responseBuffer = await response.arrayBuffer();
+
+  // TODO(#226): Migrate to Promise-syntax once Safari supports it.
+  return new Promise((resolve, reject) => {
+    context.decodeAudioData(responseBuffer, resolve, reject);
+  });
+}
+
+class Kit {
+  constructor(id, prettyName, index) {
+    this.id = id;
+    this.prettyName = prettyName;
+    this.index = index;
+    this.buffer = {};
+  }
+
+  getSampleUrl(instrumentName) {
+    return `sounds/drum-samples/${this.id}/${instrumentName.toLowerCase()}.wav`;
+  }
+
+  load() {
+    const instrumentPromises = INSTRUMENTS.map(
+        (instrument) => this.loadSample(instrument.name));
+    const promise = Promise.all(instrumentPromises).then(() => null);
+    // Return original Promise on subsequent load calls to avoid duplicate
+    // loads.
+    this.load = () => promise;
+    return promise;
+  }
+
+  async loadSample(instrumentName) {
+    this.buffer[instrumentName] = await fetchAudio(
+        this.getSampleUrl(instrumentName));
+  }
+}
+
+
+class Effect {
+  constructor(data, index) {
+    this.name = data.name;
+    this.url = data.url;
+    this.dryMix = data.dryMix;
+    this.wetMix = data.wetMix;
+    this.index = index;
+    this.buffer = undefined;
+  }
+
+  async load() {
+    // Return if buffer has been loaded already or there is nothing to load
+    // ("No effect" instance).
+    if (!this.url || this.buffer) {
+      return;
+    }
+
+    this.buffer = await fetchAudio(this.url);
+  }
+}
+
+
+class Beat {
+  constructor(data, kits, effects) {
+    this.kits = kits;
+    this.effects = effects;
+    this.loadObject(data);
+  }
+
+  loadObject(data) {
+    this._data = Object.seal(clone(data));
+    this._kit = this.kits[data.kitIndex];
+    this._effect = this.effects[data.effectIndex];
+  }
+
+  toObject() {
+    return clone(this._data);
+  }
+
+  set kit(kit) {
+    this._kit = kit;
+    this._data.kitIndex = kit.index;
+  }
+
+  get kit() {
+    return this._kit;
+  }
+
+  set effect(effect) {
+    this._effect = effect;
+    this._data.effectIndex = effect.index;
+
+    // Hack - if effect is turned all the way down - turn up effect slider.
```

Comment:
I couldn't understand from L105~L109. Can you explain why these hacks are needed?

Reply:
This code is copied from the original source and I kept the behavior from the original drum machine: when a user silences an effect by dragging the dry/wet slider to 0 and then selects a new effect from the dropdown, this code makes sure the new effect is audible by resetting the dry/wet slider to 0.5.

Reply:
Thanks for the explanation! SGTM.